webapp/src/Entity/Problem.php

```diff
@@ -287,26 +292,84 @@ public function getSpecialCompareArgs(): ?string
         return $this->special_compare_args;
     }
 
-    public function setCombinedRunCompare(bool $combinedRunCompare): Problem
+    public function setTypesAsString(array $types): Problem
     {
-        $this->combined_run_compare = $combinedRunCompare;
+        $stringToTypes = array_flip($this->typesToString);
+        $typeConstants = [];
+        foreach ($types as $type) {
+            if (!isset($stringToTypes[$type])) {
+                throw new Exception("Unknown problem type: '$type', must be one of " . implode(', ', array_keys($stringToTypes)));
+            }
+            $typeConstants[$type] = $stringToTypes[$type];
+        }
+        $this->setTypes($typeConstants);
+
         return $this;
     }
 
-    public function getCombinedRunCompare(): bool
+    public function getTypesAsString(): string
+    {
+        $typeConstants = $this->getTypes();
+        $typeStrings = [];
+        foreach ($typeConstants as $type) {
+            if (!isset($this->typesToString[$type])) {
+                throw new Exception("Unknown problem type: '$type'");
+            }
+            $typeStrings[] = $this->typesToString[$type];
+        }
+        return implode(', ', $typeStrings);
+    }
+
+    public function getTypes(): array
     {
-        return $this->combined_run_compare;
+        $ret = [];
+        foreach (array_keys($this->typesToString) as $type) {
+            if ($this->types & $type) {
+                $ret[] = $type;
+            }
+        }
+        return $ret;
     }
 
-    public function setMultipassProblem(bool $isMultipassProblem): Problem
+    public function setTypes(array $types): Problem
     {
-        $this->isMultipassProblem = $isMultipassProblem;
+        $types = array_unique($types);
+        $this->types = 0;
+        foreach ($types as $type) {
+            $this->types |= $type;
+        }
+        if (!($this->types & self::TYPE_PASS_FAIL) xor ($this->types & self::TYPE_SCORING)) {
+            throw new Exception("Invalid problem type: must be exactly one of 'pass-fail' or 'scoring'.");
+        }
```

Comment:
Note that `pass-fail` is implied when it is not provided. For example, `type: interactive` implies `type: [pass-fail, interactive]`. I noticed this when doing a test upload today :joy: 

See also: https://github.com/Kattis/problem-package-format/issues/438